File: ButtonCore/src/main/java/buttondevteam/lib/TBMCChatEvent.java
package buttondevteam.lib;

import buttondevteam.component.channel.Channel;
import buttondevteam.lib.chat.ChatMessage;
import lombok.Getter;
import lombok.experimental.Delegate;
import org.bukkit.command.CommandSender;
import org.bukkit.event.HandlerList;

import javax.annotation.Nullable;

/**
 * Make sure to only send the message to users where {@link #shouldSendTo(CommandSender)} returns true.
 * 
 * @author NorbiPeti
 *
 */
@Getter
public class TBMCChatEvent extends TBMCChatEventBase {
	public TBMCChatEvent(Channel channel, ChatMessage cm, Channel.RecipientTestResult rtr) {
		super(channel, cm.getMessage(), rtr.score, rtr.groupID);
		this.cm = cm;
    }

	private static final HandlerList handlers = new HandlerList();

	@Delegate //<-- Backwards compatibility
	private ChatMessage cm;

	private boolean isIgnoreSenderPermissions() {
		return cm.getPermCheck() != cm.getSender();
	}
	// TODO: Message object with data?

    /**
     * This will allow the sender of the message if {@link #isIgnoreSenderPermissions()} is true.
     */
    @Override
    public boolean shouldSendTo(CommandSender sender) {
	    if (isIgnoreSenderPermissions() && sender.equals(this.cm.getSender()))
            return true; //Allow sending the message no matter what
        return super.shouldSendTo(sender);
    }

    /**
     * This will allow the sender of the message if {@link #isIgnoreSenderPermissions()} is true.
     */
    @Override
    public int getMCScore(CommandSender sender) {
	    if (isIgnoreSenderPermissions() && sender.equals(this.cm.getSender()))
            return getScore(); //Send in the correct group no matter what
        return super.getMCScore(sender);
    }

    /**
     * This will allow the sender of the message if {@link #isIgnoreSenderPermissions()} is true.
     */
    @Nullable
    @Override
    public String getGroupID(CommandSender sender) {
	    if (isIgnoreSenderPermissions() && sender.equals(this.cm.getSender()))
            return getGroupID(); //Send in the correct group no matter what
        return super.getGroupID(sender);
    }

    @Override
	public HandlerList getHandlers() {
		return handlers;
	}

	public static HandlerList getHandlerList() {
		return handlers;
	}
}

File: Chroma-Core/src/main/java/buttondevteam/lib/TBMCChatEvent.kt
package buttondevteam.lib

import buttondevteam.core.component.channel.Channel
import buttondevteam.core.component.channel.Channel.RecipientTestResult
import buttondevteam.lib.chat.ChatMessage
import buttondevteam.lib.player.ChromaGamerBase
import org.bukkit.event.HandlerList

/**
 * Make sure to only send the message to users where [.shouldSendTo] returns true.
 *
 * @author NorbiPeti
 */
class TBMCChatEvent(
    channel: Channel,
    val chatMessage: ChatMessage,
    rtr: RecipientTestResult
) : TBMCChatEventBase(channel, chatMessage.message, rtr.score, rtr.groupID!!) {

    private val isIgnoreSenderPermissions: Boolean get() = chatMessage.permCheck !== chatMessage.user

    /**
     * This will allow the sender of the message if [.isIgnoreSenderPermissions] is true.
     */
    override fun shouldSendTo(sender: ChromaGamerBase): Boolean {
        return if (isIgnoreSenderPermissions && sender == chatMessage.user) true else super.shouldSendTo(sender) //Allow sending the message no matter what
    }

    /**
     * This will allow the sender of the message if [.isIgnoreSenderPermissions] is true.
     */
    override fun getMCScore(sender: ChromaGamerBase): Int {
        return if (isIgnoreSenderPermissions && sender == chatMessage.user) score else super.getMCScore(sender) //Send in the correct group no matter what
    }

    /**
     * This will allow the sender of the message if [.isIgnoreSenderPermissions] is true.
     */
    override fun getGroupID(sender: ChromaGamerBase): String? {
        return if (isIgnoreSenderPermissions && sender == chatMessage.user) groupID else super.getGroupID(sender) //Send in the correct group no matter what
    }

    override fun getHandlers(): HandlerList {
        return handlerList
    }

    /** @see ChatMessage.user */
    val user: ChromaGamerBase get() = chatMessage.user

    /** @see ChatMessage.origin */
    val origin: String get() = chatMessage.origin

    /** @see ChatMessage.isFromCommand */
    val isFromCommand get() = chatMessage.isFromCommand

    companion object {
        val handlerList = HandlerList()
    }
}